### PR TITLE
8335583: Avoid using pointers in CDS tables

### PR DESCRIPTION
**Pull request by matias9927**
There are several calls to mark_pointer which can be avoided if offsets are used instead, This will help to reduce the archive size and reduce startup time.

This patch introduces some utility functions in the first commit to assist in converting to and from u4 offsets. For ease of review, consider looking at each commit separately as they each address an individual pointer to offset change. Verified with tier 1-5 tests.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8335583](https://bugs.openjdk.org/browse/JDK-8335583): Avoid using pointers in CDS tables (**Enhancement** - P4)

### Reviewers
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21837/head:pull/21837` \
`$ git checkout pull/21837`

Update a local copy of the PR: \
`$ git checkout pull/21837` \
`$ git pull https://git.openjdk.org/jdk.git pull/21837/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21837`

View PR using the GUI difftool: \
`$ git pr show -t 21837`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21837.diff">https://git.openjdk.org/jdk/pull/21837.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21837#issuecomment-2452344432)
</details>

### CODE REVIEW

File: src/hotspot/share/cds/lambdaFormInvokers.cpp [hunk 1]
```diff
@@ -239,7 +238,13 @@ void LambdaFormInvokers::dump_static_archive_invokers() {
 void LambdaFormInvokers::read_static_archive_invokers() {
   if (_static_archive_invokers != nullptr) {
     for (int i = 0; i < _static_archive_invokers->length(); i++) {
-      Array<char>* line = _static_archive_invokers->at(i);
+      u4 offset = _static_archive_invokers->at(i);
+      Array<char>* line;
+      if (ArchiveBuilder::is_active()) {
+        line = ArchiveBuilder::current()->offset_to_buffered<Array<char>*>(offset);
+      } else {
```
Comment (iklam) [inline src/hotspot/share/cds/lambdaFormInvokers.cpp:245]: The `if (ArchiveBuilder::is_active()` branch doesn't seem necessary, as this function is called in only a single place where this condition is false.

File: src/hotspot/share/cds/runTimeClassInfo.hpp [hunk 2]
```diff
@@ -173,12 +175,12 @@ class RunTimeClassInfo {
...
   }
 
-  InstanceKlass** nest_host_addr() {
-    assert(_klass->is_hidden(), "sanity");
-    return (InstanceKlass**)(address(this) + nest_host_offset());
-  }
   InstanceKlass* nest_host() {
-    return *nest_host_addr();
+    if (ArchiveBuilder::is_active() && ArchiveBuilder::current()->is_in_buffer_space((address)this)) {
+      return ArchiveBuilder::current()->offset_to_buffered<InstanceKlass*>(_nest_host_offset);
```
Comment (iklam) [inline src/hotspot/share/cds/runTimeClassInfo.hpp:180]: I think this function is never called under this condition. Maybe the `if` should be turned into an assert:

```
assert(!ArchiveBuilder::is_active(), "not called when dumping archive");
```

**Review by iklam — CHANGES_REQUESTED** [review]

**Review by iklam — APPROVED** [review]
LGTM

**Review by calvinccheung — APPROVED** [review]
Looks good.

**Pull request closed without merging**